e2e/config/azure.go

```diff
@@ -324,10 +359,105 @@ func (a *AzureClient) assignReaderRoleToBlobStorage(ctx context.Context, princip
 		if errors.As(err, &respError) && respError.StatusCode == http.StatusConflict {
 			return nil
 		}
-		return fmt.Errorf("assign reader role: %w", err)
+		return fmt.Errorf("assign Storage Blob Data Contributor role: %w", err)
 	}
 	return nil
+}
+
+func (a *AzureClient) LatestSIGImageVersionByTag(ctx context.Context, image *Image, tagName, tagValue string) (VHDResourceID, error) {
+	galleryImageVersion, err := armcompute.NewGalleryImageVersionsClient(image.Gallery.SubscriptionID, a.Credential, a.ArmOptions)
+	if err != nil {
+		return "", fmt.Errorf("create a new images client: %v", err)
+	}
+	pager := galleryImageVersion.NewListByGalleryImagePager(image.Gallery.ResourceGroup, image.Gallery.Name, image.Name, nil)
+	var latestVersion *armcompute.GalleryImageVersion
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return "", fmt.Errorf("failed to get next page: %w", err)
+		}
+		versions := page.Value
+		for _, version := range versions {
+			// skip images tagged with the no-selection tag, indicating they
+			// shouldn't be selected dynmically for running abe2e scenarios
+			if _, ok := version.Tags[noSelectionTagName]; ok {
+				continue
+			}
+			if tagName != "" {
+				tag, ok := version.Tags[tagName]
+				if !ok || tag == nil || *tag != tagValue {
+					continue
+				}
+			}
+
+			if err := ensureProvisioningState(version); err != nil {
+				continue
+			}
+			if latestVersion == nil || version.Properties.PublishingProfile.PublishedDate.After(*latestVersion.Properties.PublishingProfile.PublishedDate) {
+				latestVersion = version
+			}
+		}
+	}
+	if latestVersion == nil {
+		return "", ErrNotFound
+	}
+
+	if err := a.ensureReplication(ctx, image, latestVersion); err != nil {
+		return "", fmt.Errorf("ensuring image replication: %w", err)
+	}
 
+	return VHDResourceID(*latestVersion.ID), nil
+}
+
+func (a *AzureClient) ensureReplication(ctx context.Context, image *Image, version *armcompute.GalleryImageVersion) error {
+	if replicatedToCurrentRegion(version) {
+		return nil
+	}
+	return a.replicateToCurrentRegion(ctx, image, version)
+}
+
+func (a *AzureClient) replicateToCurrentRegion(ctx context.Context, image *Image, version *armcompute.GalleryImageVersion) error {
+	galleryImageVersion, err := armcompute.NewGalleryImageVersionsClient(image.Gallery.SubscriptionID, a.Credential, a.ArmOptions)
```

Comment:
why moving the Client Constructor inside the function that uses it ? while all other clients are at the top ?

Reply:
I'm guessing this is to support passing different SubID, compared to the global client which are configured once ?

Reply:
Yes. Unlike others, subscription id has to me dynamic for this function.